from setuptools import setup

with open('README.md') as f:
    long_description = f.read()

setup(name='vcontact2',
      version='0.11.2',
      description='Viral Contig Automatic Clustering and Taxonomy',
      url='https://bitbucket.org/MAVERIClab/vcontact2',
      author='Benjamin Bolduc',
      long_description=long_description,
      long_description_content_type='text/markdown',
      author_email='bolduc.10@osu.edu',
      license='GPLv3',
      packages=['vcontact2', 'vcontact2.exports', 'vcontact2.utilities'],
      package_data={'vcontact2': ['data/ViralRefSeq-prokaryotes-v88.faa.gz',
                                  'data/ViralRefSeq-prokaryotes-v88.protein2contig.csv',
                                  'data/ViralRefSeq-prokaryotes-v88.Merged-reference.csv',
                                  'data/ViralRefSeq-prokaryotes-v85.faa.gz',
                                  'data/ViralRefSeq-prokaryotes-v85.protein2contig.csv',
                                  'data/ViralRefSeq-prokaryotes-v85.ICTV-reference.csv',
                                  'data/ViralRefSeq-prokaryotes-v85.Merged-reference.csv',
                                  'data/ViralRefSeq-archaea-v85.faa.gz',
                                  'data/ViralRefSeq-archaea-v85.protein2contig.csv',
                                  'data/ViralRefSeq-archaea-v85.Merged-reference.csv',
                                  'data/ViralRefSeq-prokaryotes-v94.faa.gz',
                                  'data/ViralRefSeq-prokaryotes-v94.protein2contig.csv',
                                  'data/ViralRefSeq-prokaryotes-v94.Merged-reference.csv',
                                  'data/ViralRefSeq-archaea-v94.faa.gz',
                                  'data/ViralRefSeq-archaea-v94.protein2contig.csv',
                                  'data/ViralRefSeq-archaea-v94.Merged-reference.csv'
                                  'data/ViralRefSeq-prokaryotes-v97.faa.gz',
                                  'data/ViralRefSeq-prokaryotes-v97.protein2contig.csv',
                                  'data/ViralRefSeq-prokaryotes-v97.Merged-reference.csv',
                                  'data/ViralRefSeq-archaea-v97.faa.gz',
                                  'data/ViralRefSeq-archaea-v97.protein2contig.csv',
                                  'data/ViralRefSeq-archaea-v97.Merged-reference.csv',
                                  'data/ViralRefSeq-prokaryotes-v201.faa.gz',
                                  'data/ViralRefSeq-prokaryotes-v201.protein2contig.csv',
                                  'data/ViralRefSeq-prokaryotes-v201.Merged-reference.csv',
                                  'data/ViralRefSeq-archaea-v201.faa.gz',
                                  'data/ViralRefSeq-archaea-v201.protein2contig.csv',
                                  'data/ViralRefSeq-archaea-v201.Merged-reference.csv',
                                  'data/ViralRefSeq-prokaryotes-v207.faa.gz',
                                  'data/ViralRefSeq-prokaryotes-v207.protein2contig.csv',
                                  'data/ViralRefSeq-prokaryotes-v207.Merged-reference.csv',
                                  'data/ViralRefSeq-archaea-v207.faa.gz',
                                  'data/ViralRefSeq-archaea-v207.protein2contig.csv',
                                  'data/ViralRefSeq-archaea-v207.Merged-reference.csv',
                                  'data/ViralRefSeq-prokaryotes-v211.faa.gz',
                                  'data/ViralRefSeq-prokaryotes-v211.protein2contig.csv',
                                  'data/ViralRefSeq-prokaryotes-v211.Merged-reference.csv',
                                  'data/ViralRefSeq-archaea-v211.faa.gz',
                                  'data/ViralRefSeq-archaea-v211.protein2contig.csv',
                                  'data/ViralRefSeq-archaea-v211.Merged-reference.csv'
                                  ]},
      scripts=['bin/vcontact2', 'bin/vcontact2_gene2genome'],
      install_requires=[
        'networkx>=2.2',
        'numpy>=1.20.1',
        'scipy>=1.6.0',
        'pandas>=1.0.5',
        'scikit-learn>=0.24.1',
        'biopython>=1.78',
        'tables>=3.4.0',
        'pyparsing>=2.4.6',
        'clusterone',  # Bioconda=1.0 only, pip has a lot
        # 'diamond>=0.9.14', conda-forge has different diamond
        'psutil>=5.8.0'
      ]
      )
